SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentReader.cpp

```diff
@@ -110,11 +110,18 @@ void ComponentReader::ApplyComponentData(const Worker_ComponentId ComponentId, S
 	Schema_Object* ComponentObject = Schema_GetComponentDataFields(Data);
 
 	// ComponentData will be missing fields if they are completely empty (options, lists, and maps).
-	// However, we still want to apply this empty data, so we need the full list of field IDs for
-	// that component type (Data, OwnerOnly, Handover, etc.).
-	const TArray<Schema_FieldId>& InitialIds = ClassInfoManager->GetFieldIdsByComponentId(ComponentId);
+	// Additional they may be empty if they are currently not active and not being replicated.
+	TArray<Schema_FieldId> UpdatedIds;
+	UpdatedIds.SetNumUninitialized(Schema_GetUniqueFieldIdCount(ComponentObject));
+	Schema_GetUniqueFieldIds(ComponentObject, UpdatedIds.GetData());
+	const TArray<Schema_FieldId>& ListIDs = ClassInfoManager->GetListIdsByComponentId(ComponentId);
+	for (Schema_FieldId ID : ListIDs)
+	{
+		// The list might be out of order because of these adds
+		UpdatedIds.AddUnique(ID);
```

Comment:
This makes reading a component quadratic in the number of fields. Would it be better to just apply the IDs received and then have a separate run to check the lists IDs and only do something if the specific count for that ID is 0?

Reply:
I am not sure I totally understand what you are suggesting?

Reply:
There is a nicer algorithm we can use with linear complexity actually

Reply:
Discussed a lot offline. Without a benchamrk the n log n complexity of sorting and deduplicating is probably fine. We suspect that both lists are actually sorted by default, so if we can use that then we can do this linearly if we need to. We can also reduce allocation if we need to be, but as said, we don't have numbers yet.

Reply:
If you refresh the page you'll see a slightly more optimal version.